core/src/main/scala/com/evolutiongaming/kafka/flow/PartitionFlow.scala

```diff
@@ -283,9 +287,31 @@ object PartitionFlow {
     }
 
     val release: F[Unit] = if (config.commitOnRevoke) {
-      offsetToCommit.flatMap { offset =>
-        offset.traverse_ { offset =>
-          log.info(s"committing on revoke: $offset") *> scheduleCommit.schedule(offset)
+
+      // Outside F is about collecting handles to the offsets held by existing cache entries;
+      // Inside F is about calling and aggregating those handles, which can be done even after cache was cleared;
+      def offsetsHeldByCurrentKeys: F[F[Option[Offset]]] =
+        cache.values1.flatMap { map =>
+          map
+            .toSeq
+            .traverse {
+              case (_, Right(value)) => value.context.holding.pure[F]
+              case (key, Left(_)) =>
+                log
+                  .error(s"trying to compute offset to commit but value for key $key is not ready")
+                  .as(none[Offset].pure[F])
+            }
+            .map(_.foldMapM(identity)(Async[F], pickMinOffset))
+        }
+
+      // We first gather "holding" handles for existing keys, then we clear the cache, forcing the entries to flush
+      // the state and remove the offset hold if they are configured to do that,
```

Comment:
let's mention that this is applicable to using `TimerFlowOf#flushOnCancel` and/or other strategies with `flushOnRevoke` to make the connection between them and this more obvious

I also think it's important to elaborate on this specific case even further, mentioning that after all handles flush the state and remove their currently held offset (there will be none), the offset to commit will correspond to the offset of the latest incoming batch of records

ideally, I would prefer it to describe both cases (using/not using `flushOnCancel`) and what offset is expected to be calculated here. i'm asking this because the relation between this code and everything happening in `TimerFlowOf` is very opaque and obscure (even though those are tightly coupled), requires building the whole mental model of code execution in one's head and confuses even me every time i spend too much time away from this code. it would greatly help everyone contributing to this code in the future

Reply:
good point, will add an extended comment

Reply:
done